.github/workflows/arduino_test_runner.yml

```diff
@@ -0,0 +1,17 @@
+name: Arduino CI
+
+on: [push, pull_request]
+
+jobs:
+  runTest:
+    runs-on: ubuntu-latest
+    timeout-minutes: 20
+
+    steps:
+      - uses: actions/checkout@v4
+      - uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: 2.6
+      - run: |
+          gem install arduino_ci
+          arduino_ci.rb
```

Comment:
We can probably read up on it on the documentation [Arduino-CI](https://github.com/Arduino-CI/arduino_ci), but let us also ask right here: What does invoking `arduino_ci.rb` actually do in this context? Does it run the sketch, and how?

On the Arduino-CI docs, I can find:

> [...] provides a system of mocks that allow fine-grained control over the hardware inputs, including the system's clock

We can't imagine some set of standard mocks can satisfy the requirements of HaniMandl easily. What about the display interface, for example?

Is there a reason attached to this, that the code has been duplicated into `examples/hanimandl/hanimandl.ino`?

Reply:
We guess the current state "just compiles" the code, and verifies it on behalf of that outcome, just like [platformio-ci.yaml](https://github.com/hiveeyes/hanimandl/blob/main/.github/workflows/platformio-ci.yaml) is doing it, right?

Coming from there, do you have any suggestions how we could apply better unit testing to the codebase? I guess it would need to be tremendously modularized, in order to achieve that?

_Apologies for our beginner's questions. Please do not spend too much time answering them, we are also writing them down for our own reference, in order to outline where we will need to level up on understanding and using Arduino-CI properly. The right way will be to do it hands-on, paired with reading relevant documentation sections. We just didn't get around doing it, yet._

Reply:
> What does invoking arduino_ci.rb actually do in this context? Does it even "run" the sketch to completion, and how?

arduino_ci.rb  is a ruby script that calls the command line Arduino compilers.

Reply:
>  We can't imagine some set of standard mocks can satisfy the requirements of HaniMandl easily. What about the display interface, for example?

Mocks are not too difficult, but takes quite some effort to implement, so I do not know if it is worth the effort.

Mocks are ideal for testing a part of a library, when you would create some core "hanimandl" library it might make sense to put the effort in. A way to do this is to split of typical functions in a separate ```hanimadl.h``` file and then you would be able to test these functions.

Reply:
> We guess the current state "just compiles" the code, and verifies it on behalf of that outcome, just like [platformio-ci.yaml](https://github.com/hiveeyes/hanimandl/blob/main/.github/workflows/platformio-ci.yaml) is doing it, right?

That is right, however it does compile the code under the Arduino IDE, not under the PlatformIO environment. As these are different it adds "robustness". The same is true for doing the LINT check, it does look at the code and even if it compiles it warns you of some risky code constructs.

Reply:
> Coming from there, do you have any suggestions how we could apply better unit testing to the codebase? I guess it would need to be tremendously modularized, in order to achieve that?

Yes, think so.
As said before, the first step is to extract some typical hanimandl functions.
 
I looked quickly at the code and I see you have organized it in a way that keeps all things close to each other, however many functions do handle pins, do display work and changes some setting etc. A program of this size I would split of all display handling in one file, all the reading of buttons (pin IO) in one file, the handling of rotary encoders in one file.
Furthermore I would have a STATE variable that defines the modus operandi of the program. 
Am I in state_setup, state_filling or state_error, etc. 
So yes, quite some work.